### ISSUE TITLE
Need help to solve the error

### ISSUE THREAD

**Issue by SinDongHwan** [gsig/temporal-fields#6]
Hello, 
I am an intern and would like to run temporal-fields package. 
(I could run Charades successfully.)
When I run temporal-fields I faced the following errors that I could not solve by myself.
Would anyone out there help me?
Thanks.
Donghwan

---------------------------------------------

warning: could not load nccl, falling back to default communication
Creating Trainer
=> Training epoch # 1
Found Environment variable CUDNN_PATH = /usr/local/cuda-8.0/lib64/libcudnn.so
Found Environment variable CUDNN_PATH = /usr/local/cuda-8.0/lib64/libcudnn.so
FATAL THREAD PANIC: (read) /home/jwkim/torch/install/share/lua/5.1/torch/File.lua:343: unknown Torch class <nn.BlockGradient>
FATAL THREAD PANIC: (read) /home/jwkim/torch/install/share/lua/5.1/torch/File.lua:343: unknown Torch class <nn.BlockGradient>
Segmentation fault (core dumped)

**Comment by gsig** [comment]
Let's try to figure his out!

What is your configuration file? (exp/*.lua) Are you training from the ImageNet pretrained model?

It is complaining about nn.BlockGradient being missing which is defined under layers/, is the directory structure intact? (This repo has more files than charades-algorithms).

Best,
Gunnar

**Comment by gsig** [comment]
You can try adding the following to models/init.lua:
require 'layers/BlockGradient'
require 'layers/Constant'

**Comment by SinDongHwan** [comment]
Thank you Gunnar,

**What is your configuration file? (exp/*.lua) Are you training from the ImageNet pretrained model?
It is complaining about nn.BlockGradient being missing which is defined under layers/, is the directory structure intact? (This repo has more files than charades-algorithms).**

Yes, I've downloaded the ImageNet pretained model from this github and installed on my linux.
I haven't change any structure under layers/

Now, I tried what you told me but still I got the same error.

_**You can try adding the following to models/init.lua:
require 'layers/BlockGradient'
require 'layers/Constant'**_
Yes, I tried to add the two lines in init.lua and could add only "require 'layers/BlockGradient'". 
This is because " require 'layers/Constant' " is already included in nn.Constants in torch, so, it makes conflicts and I couldn't add this line.
After adding the "require 'layers/BlockGradient'", I got the same error messages as before.

Any clue to solve this problem?
Donghwan.

**Comment by gsig** [comment]
Did you mean to close the issue? (Feel free to reopen it otherwise)

Can you send the contents of your configuration file? (exp/*.lua) I want to see if I can replicate this.
The fact that is is complaining about a "FATAL THREAD PANIC" suggests that the error is coming from the data loading threads (anything related to dataloader.lua or datasets/*) which is strange given that it is complaining about nn.BlockGradient.

Do you have a stack trace? It might help to change nThreads to 1 in the configuration, and use a debugger such as https://github.com/slembcke/debugger.lua when you are launching the script. That is, in your configuration file (exp/*.lua), change dofile 'main.lua' to:

local dbg = require("debugger")
dbg.call(function()
    dofile 'main.lua'
end)

Best,
Gunnar

**Comment by SinDongHwan** [comment]
Woops! I didn't intended to close the issue. By some reason, it is closed. Sorry. Please keep helping me.

My research goal is converting a video with objects and activities into a Korean natural language in text sentence. So, I would like to investigate how to extract names of objects and activities from an input video on your temporal-fields program and store them into an ontology based knowledge base. ( I am just a beginner in deep learning and lua language, so suffering many difficulties at the beginning.)

Regarding the error issue, I will follow what you told me above and will send you the config file (exp/.lua)

Again, thank you very much for your detail comments. It helps me a lot!

**Comment by gsig** [comment]
This code is able to give you a distribution over objects,verbs,actions etc.

The function CRF_ATFloss:actions() which is called by crftrain.lua converts the local object,verb,scene,goal tensor into actions by indexing the appropriate object,verb pairs. As a first approximation, you can just use the output of this [line](https://github.com/gsig/temporal-fields/blob/824712c8b2c5e682be87d106b7b3709772cd0671/crftrain.lua#L296). This output is written to a file, so you can just read that file. You can always sum up the probabilities for actions corresponding to the same object to get a distribution over the objects.

If you want to get a proper distribution over the actions you can add a function as follows (untested)
```
function CRF_ATFloss:objects(a2ov) -- call forward first
    -- Calculates the marginal of the objects
    local nO,nV,nS = 38,33,16
    local out = torch.Tensor(self.batchSize,nO)
    local o,v = unpack(a2ov[a-1]) --0-index
    for o=1,nO do
        local tmp1 = lse(self.Qi[{{},o+1,{},{},{}}],23)[{{},1,1,{}}]
        local tmp2 = lse(self.Qi[{{},o+1+nO,{},{},{}}],23)[{{},1,1,{}}]
        local tmp3 = lse(self.Qi[{{},o+1+nO*2,{},{},{}}],23)[{{},1,1,{}}]
        local gt_g_marginal = lse(tmp1:cat(tmp2,3):cat(tmp3,3), 3)
        out[{{},o}] = (lse(gt_g_marginal,2):squeeze():float()):exp()
    end
end
```

I am trying to replicate your error. Do you get the same error for both rgb.lua and rgb_test.lua?

**Comment by gsig** [comment]
Also, can you attach the whole <cachedir>/log.txt file for both rgb.lua and rgb_test.lua?

I cloned the newest repo, ran get_models.sh and get_alreadytrained.sh, edited cacheDir and data in your config files and it worked for me. I also uncommented the nGPU, because I just set `export CUDA_VISIBLE_DEVICES=2` before opening torch (to avoid allocating memory on all gpus)

What OS are you using? And what version of torch/cuda/cudnn?

**Comment by SinDongHwan** [comment]
Thank you again Gunnar,
With your comments, I could resolve the problem that I've faced.
Here I am reporting you what I did;

step1. I moved  BlockGradient.lua in "~/temporal-fields/layers" into  "~/torch/install/share/lua/5.1/nn/".
step2. Add "require('nn.BlockGradient')" in "~/torch/install/share/lua/5.1/nn/init.lua".
step3. Commented out "require('layers/BlockGradient)" in  "~/temporal-fields/layers/CRF_ATF.lua". This is done to escape from conflict of  'layers/BlockGradient.lua' and 'nn/BlockGradient.lua' 

It's working now!

Regarding your detail and kind explanation where I can look into your code to extract the names of objects, your comments help me a lot !
I will keep let you know what I do.

Have a good day, Gunnar.
Best Regards,
Dong-Hwan
